File: frontend/iosApp/UILogics/Sources/UILogics/Screens/AlbumFormViewModel.swift
import Foundation
import SwiftUI
import Domains
import UseCases

@MainActor
public final class AlbumFormViewModel: ObservableObject {
    @Published public var title: String
    @Published public var coverImage: ImageUIModel?
    @Published public var selectedImage: UIImage? {
        didSet {
            if let image = selectedImage {
                coverImage = .selectedImage(image)
            }
        }
    }
    @Published public var isShowingImagePicker = false
    @Published public var isSaving = false
    @Published public var alertItem: AlertItem?

    public let mode: AlbumFormMode
    private let useCase: AlbumFormUseCaseProtocol
    private let router: AuthenticatedRouterProtocol

    public var navigationTitle: String {
        switch mode {
        case .create:
            return "New Album"
        case .edit:
            return "Edit Album"
        }
    }

    public init(mode: AlbumFormMode, useCase: AlbumFormUseCaseProtocol, router: AuthenticatedRouterProtocol) {
        self.mode = mode
        self.useCase = useCase
        self.router = router

        switch mode {
        case .create:
            self.title = ""
            self.coverImage = nil
        case .edit(let album):
            self.title = album.title
            self.coverImage = album.displayCoverImageURL.map { .uploadedImage($0) }
        }
    }

    public var isValid: Bool {
        !title.trimmingCharacters(in: .whitespaces).isEmpty
    }

    public func selectCoverImage() {
        isShowingImagePicker = true
    }

    public func save() {
        Task {
            await saveAlbum()
        }
    }

    public func cancel() {
        router.dismissSheet()
    }

    private func saveAlbum() async {
        isSaving = true

        let imageData: Data? = {
            guard case .selectedImage(let image) = coverImage else {
                return nil
            }
            return image.jpegData(compressionQuality: 0.8)
        }()

        switch mode {
        case .create:
            await createAlbum(imageData: imageData)
        case .edit(let album):
            await updateAlbum(album: album, imageData: imageData)
        }

        isSaving = false
    }

    private func createAlbum(imageData: Data?) async {
        let result = await useCase.createAlbum(
            title: title.trimmingCharacters(in: .whitespaces),
            coverImageData: imageData
        )

        switch result {
        case .success, .successPendingSync:
            router.dismissSheet()
        case .failure(let error):
            showAlert(for: error)
        }
    }

    private func updateAlbum(album: Album, imageData: Data?) async {
        let result = await useCase.updateAlbum(
            album: album,
            title: title.trimmingCharacters(in: .whitespaces),
            coverImageData: imageData
        )

        switch result {
        case .success, .successPendingSync:
            router.dismissSheet()
        case .failure(let error):
            showAlert(for: error)
        }
    }

    private func showAlert(for error: AlbumFormUseCaseModel.CreateResult.Error) {
        let message: String
        switch error {
        case .networkError:
            message = "Network error. Please check your connection and try again."
        case .serverError:
            message = "Server error. Please try again later."
        case .imageStorageFailed:
            message = "Failed to save the image locally. Please try again."
        case .databaseError:
            message = "Failed to save to local database. Please try again."
        case .unknown:
            message = "An unexpected error occurred. Please try again."
        }

        alertItem = AlertItem(
            title: "Save Failed",
            message: message,
            buttons: [Alert.Button.default(Text("OK"))]
        )
    }

    private func showAlert(for error: AlbumFormUseCaseModel.UpdateResult.Error) {
        let message: String
        switch error {
        case .networkError:
            message = "Network error. Please check your connection and try again."
        case .serverError:
            message = "Server error. Please try again later."
        case .notFound:
            message = "Album not found. It may have been deleted."
        case .imageStorageFailed:
            message = "Failed to save the image locally. Please try again."
        case .databaseError:
            message = "Failed to save to local database. Please try again."
        case .unknown:
            message = "An unexpected error occurred. Please try again."
        }

        alertItem = AlertItem(
            title: "Save Failed",
            message: message,
            buttons: [Alert.Button.default(Text("OK"))]
        )
    }
}

// MARK: - UI Models

extension AlbumFormViewModel {
    public enum ImageUIModel: Equatable {
        case uploadedImage(URL)
        case selectedImage(UIImage)

        public var uploadedImageUrl: URL? {
            guard case let .uploadedImage(url) = self else {
                return nil
            }
            return url
        }

        public static func == (lhs: ImageUIModel, rhs: ImageUIModel) -> Bool {
            switch (lhs, rhs) {
            case (.uploadedImage(let lhsUrl), .uploadedImage(let rhsUrl)):
                return lhsUrl == rhsUrl
            case (.selectedImage(let lhsImage), .selectedImage(let rhsImage)):
                return lhsImage === rhsImage
            default:
                return false
            }
        }
    }
}

File: frontend/composeApp/src/androidMain/kotlin/com/example/memoriesapp/ui/uilogics/viewmodels/AlbumFormViewModel.kt
package com.example.memoriesapp.ui.uilogics.viewmodels

import android.graphics.Bitmap
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.memoriesapp.domain.Album
import com.example.memoriesapp.usecase.AlbumCreateResult
import com.example.memoriesapp.usecase.AlbumFormUseCase
import com.example.memoriesapp.usecase.AlbumUpdateResult
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch
import java.io.ByteArrayOutputStream

sealed class AlbumFormMode {
    data object Create : AlbumFormMode()
    data class Edit(val album: Album) : AlbumFormMode()
}

class AlbumFormViewModel(
    private val mode: AlbumFormMode,
    private val albumFormUseCase: AlbumFormUseCase
) : ViewModel() {

    var title by mutableStateOf("")
        private set
    var coverImage by mutableStateOf<CoverImageState>(CoverImageState.None)
        private set
    var isSaving by mutableStateOf(false)
        private set
    var errorMessage by mutableStateOf<String?>(null)
        private set

    private val _navigationEvent = MutableSharedFlow<NavigationEvent>()
    val navigationEvent: SharedFlow<NavigationEvent> = _navigationEvent.asSharedFlow()

    val navigationTitle: String
        get() = when (mode) {
            is AlbumFormMode.Create -> "New Album"
            is AlbumFormMode.Edit -> "Edit Album"
        }

    val isValid: Boolean
        get() = title.trim().isNotEmpty()

    init {
        when (mode) {
            is AlbumFormMode.Create -> {
                title = ""
                coverImage = CoverImageState.None
            }
            is AlbumFormMode.Edit -> {
                title = mode.album.title
                coverImage = mode.album.displayCoverImage?.let {
                    CoverImageState.Uploaded(it)
                } ?: CoverImageState.None
            }
        }
    }

    fun onTitleChange(newTitle: String) {
        title = newTitle
    }

    fun onImageSelected(bitmap: Bitmap) {
        coverImage = CoverImageState.Selected(bitmap)
    }

    fun save() {
        viewModelScope.launch {
            saveAlbum()
        }
    }

    fun cancel() {
        viewModelScope.launch {
            _navigationEvent.emit(NavigationEvent.Dismiss(updatedAlbum = null))
        }
    }

    fun clearError() {
        errorMessage = null
    }

    private suspend fun saveAlbum() {
        isSaving = true
        errorMessage = null

        val imageData: ByteArray? = when (val image = coverImage) {
            is CoverImageState.Selected -> {
                val stream = ByteArrayOutputStream()
                image.bitmap.compress(Bitmap.CompressFormat.JPEG, 80, stream)
                stream.toByteArray()
            }
            else -> null
        }

        when (mode) {
            is AlbumFormMode.Create -> createAlbum(imageData)
            is AlbumFormMode.Edit -> updateAlbum(mode.album, imageData)
        }

        isSaving = false
    }

    private suspend fun createAlbum(imageData: ByteArray?) {
        when (val result = albumFormUseCase.createAlbum(title.trim(), imageData)) {
            is AlbumCreateResult.Success,
            is AlbumCreateResult.SuccessPendingSync -> {
                _navigationEvent.emit(NavigationEvent.Dismiss(updatedAlbum = null))
            }
            is AlbumCreateResult.Failure -> {
                errorMessage = mapCreateError(result.error)
            }
        }
    }

    private suspend fun updateAlbum(album: Album, imageData: ByteArray?) {
        when (val result = albumFormUseCase.updateAlbum(album, title.trim(), imageData)) {
            is AlbumUpdateResult.Success -> {
                _navigationEvent.emit(NavigationEvent.Dismiss(updatedAlbum = result.album))
            }
            is AlbumUpdateResult.SuccessPendingSync -> {
                _navigationEvent.emit(NavigationEvent.Dismiss(updatedAlbum = result.album))
            }
            is AlbumUpdateResult.Failure -> {
                errorMessage = mapUpdateError(result.error)
            }
        }
    }

    private fun mapCreateError(error: com.example.memoriesapp.usecase.AlbumCreateError): String {
        return when (error) {
            com.example.memoriesapp.usecase.AlbumCreateError.NETWORK_ERROR ->
                "Network error. Please check your connection and try again."
            com.example.memoriesapp.usecase.AlbumCreateError.SERVER_ERROR ->
                "Server error. Please try again later."
            com.example.memoriesapp.usecase.AlbumCreateError.IMAGE_STORAGE_FAILED ->
                "Failed to save the image locally. Please try again."
            com.example.memoriesapp.usecase.AlbumCreateError.DATABASE_ERROR ->
                "Failed to save to local database. Please try again."
            else -> "An unexpected error occurred. Please try again."
        }
    }

    private fun mapUpdateError(error: com.example.memoriesapp.usecase.AlbumUpdateError): String {
        return when (error) {
            com.example.memoriesapp.usecase.AlbumUpdateError.NETWORK_ERROR ->
                "Network error. Please check your connection and try again."
            com.example.memoriesapp.usecase.AlbumUpdateError.SERVER_ERROR ->
                "Server error. Please try again later."
            com.example.memoriesapp.usecase.AlbumUpdateError.NOT_FOUND ->
                "Album not found. It may have been deleted."
            com.example.memoriesapp.usecase.AlbumUpdateError.IMAGE_STORAGE_FAILED ->
                "Failed to save the image locally. Please try again."
            com.example.memoriesapp.usecase.AlbumUpdateError.DATABASE_ERROR ->
                "Failed to save to local database. Please try again."
            else -> "An unexpected error occurred. Please try again."
        }
    }

    sealed class CoverImageState {
        data object None : CoverImageState()
        data class Uploaded(val url: String) : CoverImageState()
        data class Selected(val bitmap: Bitmap) : CoverImageState()
    }

    sealed class NavigationEvent {
        data class Dismiss(val updatedAlbum: Album? = null) : NavigationEvent()
    }
}
